### PR TITLE
ADR 019 password rotation

### PR DESCRIPTION
**Pull request by RichardIrons-neo4j**
This PR updates the preview feature "re-auth" significantly. The changes allow for catering to a wider range of use cases including simple password rotation.

As part of this PR, all auth-related namespaces have been moved to preview - previously some did not have this, although the classes therein would not have been usable.

Since this is a preview feature all changes here are breaking changes.

The `OnTokenExpiredAsync` method in the `IAuthTokenManager` interface was removed, and a new `HandleSecurityExceptionAsync` method was added in its place.

The `ExpirationBased` method in `AuthTokenManagers` was renamed to `Bearer`, and a new `Basic` method was added to cater for password rotation.

### CODE REVIEW

File: Neo4j.Driver/Neo4j.Driver/Internal/Connector/SocketConnection.cs [hunk 1]
```diff
@@ -373,10 +375,12 @@ public async Task ValidateCredsAsync()
             token = await AuthTokenManager.GetTokenAsync().ConfigureAwait(false);
         }
 
-        // The token has changed or the connection needs to re-authenticate but can use the same credentials.
-        if (AuthorizationStatus == AuthorizationStatus.AuthorizationExpired || !token.Equals(AuthToken))
+        var authExpired = AuthorizationStatus == AuthorizationStatus.AuthorizationExpired;
+        if (authExpired || !token.Equals(AuthToken))
```
Comment (AndyHeap-NeoTech) [inline Neo4j.Driver/Neo4j.Driver/Internal/Connector/SocketConnection.cs:387]: Is there a situation here where the code would need to handle a null token being supplied from the AuthTokenManager. Previous code threw an exception in that case (see line 354 for conditional on null check).

**Review by AndyHeap-NeoTech — APPROVED** [review]
I had one question about handling a potential null token. All looks good to me with nothing obvious standing out.

File: Neo4j.Driver/Neo4j.Driver/Neo4j.Driver.csproj [hunk 2]
```diff
@@ -39,21 +39,21 @@
         <WarningLevel>2</WarningLevel>
     </PropertyGroup>
     <ItemGroup Condition="'$(TargetFramework)' == 'net5.0'">
-        <PackageReference Include="System.Runtime.InteropServices.RuntimeInformation" Version="4.3.0"/>
```
Comment (thelonelyvulpes) [inline Neo4j.Driver/Neo4j.Driver/Neo4j.Driver.csproj:42]: for whatever reason our configs are not aligned on what this xml should have, we need to pick a style.

File: Neo4j.Driver/Neo4j.Driver/Internal/ConnectionPool.cs [hunk 3]
```diff
@@ -299,19 +299,13 @@ public void OnPoolMemberException(IPooledConnection connection, Exception except
...
-                }
-            }
+            connection.AuthorizationStatus = AuthorizationStatus.SecurityError;
         }
 
         if (exception is AuthorizationException)
         {
+            // if an auth exception occured the pool member, all connections in the pool that are using that token
+            // should be closed. This is because the token is now invalid and all connections using it will fail.
             foreach (var conn in _inUseConnections)
```
Comment (thelonelyvulpes) [inline Neo4j.Driver/Neo4j.Driver/Internal/ConnectionPool.cs]: do we care about setting it for idle connections, _inUseConnections are connections that another session or transaction currently has borrowed.

**Review by thelonelyvulpes — APPROVED** [review]

File: Neo4j.Driver/Neo4j.Driver/Internal/ConnectionPool.cs (hunk 3 above)
Comment (RichardIrons-neo4j) [reply Neo4j.Driver/Neo4j.Driver/Internal/ConnectionPool.cs]: I _think_ it's right to let those connections fail the next time they're attempted to be used?

Comment (thelonelyvulpes) [reply Neo4j.Driver/Neo4j.Driver/Internal/ConnectionPool.cs]: why do we want idle connections to lazily discover the auth token is expired but set in use connections as expired?

File: Neo4j.Driver/Neo4j.Driver/Internal/Connector/SocketConnection.cs (hunk 1 above)
Comment (RichardIrons-neo4j) [reply Neo4j.Driver/Neo4j.Driver/Internal/Connector/SocketConnection.cs:387]: Added a null check.

File: Neo4j.Driver/Neo4j.Driver/Internal/ConnectionPool.cs (hunk 3 above)
Comment (RichardIrons-neo4j) [reply Neo4j.Driver/Neo4j.Driver/Internal/ConnectionPool.cs]: Yeah, thinking about it, you're right, let me alter that.

**Review by robsdedude — COMMENTED** [review]
I'm a bit late to the party, but please note that the TestKit tests for this feature introduced a new feature flag `API_RETRYABLE_EXCEPTION = "Feature:API:RetryableExceptions"`. For better test coverage, I highly recommend implementing it in the backend.

**Pull request merged**